state-chain/runtime/src/chainflip/vault_swap.rs

```diff
@@ -0,0 +1,232 @@
+					.try_into()
+					.map_err(|_| pallet_cf_swapping::Error::<Runtime>::AffiliateFeeTooHigh)?,
+			})
+		})
+		.collect::<Result<Vec<AffiliateAndFee>, _>>()?;
+
+	<BoundedVec<AffiliateAndFee, MaxAffiliates>>::try_from(affiliates_and_fees)
+		.map_err(|_| DispatchErrorWithMessage::from("Too many affiliates provided"))
+}
+
+pub fn bitcoin_vault_swap(
+	broker_id: AccountId,
+	destination_asset: Asset,
+	destination_address: EncodedAddress,
+	broker_commission: BasisPoints,
+	min_output_amount: AssetAmount,
+	retry_duration: BlockNumber,
+	boost_fee: BasisPoints,
+	affiliate_fees: Affiliates<AccountId>,
+	dca_parameters: Option<DcaParameters>,
+) -> Result<VaultSwapDetails<String>, DispatchErrorWithMessage> {
+	let private_channel_id =
+		pallet_cf_swapping::BrokerPrivateBtcChannels::<Runtime>::get(&broker_id)
+			.ok_or(pallet_cf_swapping::Error::<Runtime>::NoPrivateChannelExistsForBroker)?;
+	let params = UtxoEncodedData {
+		output_asset: destination_asset,
+		output_address: destination_address,
+		parameters: SharedCfParameters {
+			retry_duration: retry_duration
+				.try_into()
+				.map_err(|_| pallet_cf_swapping::Error::<Runtime>::SwapRequestDurationTooLong)?,
+			min_output_amount,
+			number_of_chunks: dca_parameters
+				.as_ref()
+				.map(|params| params.number_of_chunks)
+				.unwrap_or(1)
+				.try_into()
+				.map_err(|_| pallet_cf_swapping::Error::<Runtime>::InvalidDcaParameters)?,
+			chunk_interval: dca_parameters
+				.as_ref()
+				.map(|params| params.chunk_interval)
+				.unwrap_or(SWAP_DELAY_BLOCKS)
+				.try_into()
+				.map_err(|_| pallet_cf_swapping::Error::<Runtime>::InvalidDcaParameters)?,
+			boost_fee: boost_fee
+				.try_into()
+				.map_err(|_| pallet_cf_swapping::Error::<Runtime>::BoostFeeTooHigh)?,
+			broker_fee: broker_commission
+				.try_into()
+				.map_err(|_| pallet_cf_swapping::Error::<Runtime>::BrokerFeeTooHigh)?,
+			affiliates: to_affiliate_and_fees::<ConstU32<BTC_MAX_AFFILIATES>>(
+				broker_id,
+				affiliate_fees,
+			)?,
+		},
+	};
+
+	Ok(VaultSwapDetails::Bitcoin {
+		nulldata_payload: encode_swap_params_in_nulldata_payload(params),
+		deposit_address: derive_btc_vault_deposit_address(private_channel_id),
+	})
+}
+
+pub fn solana_vault_swap(
+	broker_id: AccountId,
+	source_asset: Asset,
+	destination_asset: Asset,
+	destination_address: EncodedAddress,
+	broker_commission: BasisPoints,
+	extra_parameters: VaultSwapExtraParametersEncoded,
+	channel_metadata: Option<CcmChannelMetadata>,
+	boost_fee: BasisPoints,
+	affiliate_fees: Affiliates<AccountId>,
+	dca_parameters: Option<DcaParameters>,
+) -> Result<VaultSwapDetails<String>, DispatchErrorWithMessage> {
+	// Load up environment variables.
+	let api_environment = SolEnvironment::api_environment()
+		.map_err(|_| DispatchErrorWithMessage::from("Failed to load Solana API environment"))?;
+
+	let agg_key = SolEnvironment::current_agg_key()
+		.map_err(|_| DispatchErrorWithMessage::from("Failed to load Solana Agg key"))?
+		.into();
+
```

Comment:
I was talking to @martin-chainflip and he was raising the concern that during a period of time on a rotation we might be encoding transactions that will be invalid. Those transacitons will revert on chain and there is no risk of user losing funds either. Worst case scenario the user transaction reverts (basically zero cost), and that's why the API will also return the lifetime of that transaction per say, but it will be impossible for integrators to craft a valid transaction.
The period of time would be after a new key is created and before the rotation is finished and the rotation transaction is included on chain. The biggest risk there is some engines failing to broadcast. It's safe to say that it could easily go up to minutes.
I thought about it and I don't think there is any way we can reliably craft valid transactions all the time, even if we were to change the programs. However, one thing that could be done (if it's even possible) to mitigate this is to use the `SolanaBroadcaster::CurrentOnChainKey` instead of the environment one. In that way we would decrease that time to the finality of Solana (12 sec). Wdyt, is it possible? @syan095 @dandanlen

Reply:
The change would be easy to implement. 

Do we want to this this for Vault swaps only? or for all Solana transactions we craft? 
i.e. currently we have lookup function to get the current Aggkey:
`let epoch = SolanaThresholdSigner::current_key_epoch()?;
SolanaThresholdSigner::keys(epoch)`

I can replace this to use `SolanaBroadcaster::CurrentOnChainKey` instead.

Alternatively I can add one more function that returns the on chain key, and only use this for VaultSwap transactions, so nothing else is affected.

Reply:
>We will also return the expiry of that transaction. 

Do we want to return the expiry (end of current epoch I assume?) block as part of the response?
Currently the response only contain the instruction. We can easily add the expiry block.

Reply:
We can do the expiry separately, there is a ticket already: https://linear.app/chainflip/issue/PRO-1851/add-expires-at-field-for-request-swap-parameter-encoding

Reply:
> Do we want to this this for Vault swaps only? or for all Solana transactions we craft?
If we do this change we should only do it for the Solana Vault swaps. Otherwise there'll be scenarios where the aggKey used in the payload will be different to the one used to sign and that is problematic.

> The change would be easy to implement.
In that case I think this is the cleanest solution we have. Estimating the expiry time during a rotation can't really be done in an exact way but at least there will only be ~12 sec every 3 days of crafted payloads that will not be valid and there is no risk of losing funds.

Reply:
Resolving as this is being addressed in https://github.com/chainflip-io/chainflip-backend/pull/5512